cpp/arcticdb/util/reliable_storage_lock.tpp

```diff
@@ -0,0 +1,233 @@
+    for (auto epoch : epochs) {
+        auto lock_key = RefKey{get_stream_id(epoch), KeyType::SLOW_LOCK};
+        if (get_expiration(lock_key) + timeout_ < now) {
+            to_delete.emplace_back(lock_key);
+        }
+    }
+    store_->remove_keys_sync(to_delete);
+}
+
+template <class ClockType>
+std::optional<Epoch> ReliableStorageLock<ClockType>::try_take_lock() const {
+    auto [existing_locks, latest] = get_all_locks();
+    if (latest.has_value()) {
+        auto expires = get_expiration(RefKey{get_stream_id(latest.value()), KeyType::SLOW_LOCK});
+        if (expires > ClockType::nanos_since_epoch()) {
+            // An unexpired lock exists
+            return std::nullopt;
+        }
+    }
+    return try_take_next_epoch(existing_locks, latest);
+}
+
+template<class ClockType>
+Epoch ReliableStorageLock<ClockType>::retry_until_take_lock() const {
+    // We don't use the ExponentialBackoff because we want to be able to wait indefinitely
+    auto max_wait = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::nanoseconds(timeout()));
+    auto min_wait = max_wait / 16;
+    auto current_wait = min_wait;
+    std::minstd_rand gen(std::random_device{}());
+    std::uniform_real_distribution<> dist(1.0, 1.2);
+    auto jittered_wait = [&]() {
+        auto factor = dist(gen);
+        return current_wait * factor;
+    };
+
+    auto aquired_epoch = try_take_lock();
+    while (!aquired_epoch.has_value()) {
+        std::this_thread::sleep_for(jittered_wait());
+        current_wait = std::min(current_wait * 2, max_wait);
+        aquired_epoch = try_take_lock();
+    }
+    return aquired_epoch.value();
+}
+
+template <class ClockType>
+std::optional<Epoch> ReliableStorageLock<ClockType>::try_extend_lock(Epoch held_lock_epoch) const {
+    auto [existing_locks, latest] = get_all_locks();
+    util::check(latest.has_value() && latest.value() >= held_lock_epoch,
+                "We are trying to extend a newer epoch than the existing one in storage. Extend epoch: {}",
+                held_lock_epoch);
+    if (latest.value() != held_lock_epoch) {
+        // We have lost the lock while holding it (most likely due to timeout).
+        return std::nullopt;
+    }
+    return try_take_next_epoch(existing_locks, latest);
+}
+
+template <class ClockType>
+void ReliableStorageLock<ClockType>::free_lock(Epoch held_lock_epoch) const {
+    auto [existing_locks, latest_epoch] = get_all_locks();
+    util::check(latest_epoch.has_value() && latest_epoch.value() >= held_lock_epoch,
+                "We are trying to free a newer epoch than the existing one in storage. Free epoch: {}, Existing epoch: {}",
+                held_lock_epoch, latest_epoch);
+    if (latest_epoch.value() != held_lock_epoch) {
+        // Lock is already lost
+        return;
+    }
+    auto lock_stream_id = get_stream_id(held_lock_epoch);
+    auto expiration = ClockType::nanos_since_epoch(); // Write current time to mark lock as expired as of now
```

Comment:
Yeah you're right. The difference is not big though. Just 2x the clock skew (if we e.g. expect clock skew to be 100ms for a process this would enstill a 200ms total skew between the two processes, otherwise if we use the s3 time as a source of truth we would suffer only the 100ms skew).

Let's do this together with the If-Match rework (which will also require passing in the ETag from s3 to the `Store` API and we can do a similar change to use the last-modified time)

Reply:
I think that's reasonable on the basis that clock drift can still cause locking failures even if we make the change I suggested.